src/logic/hydra.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2022 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import data.multiset.basic
+import order.well_founded
+
+/-!
+# Termination of a hydra game
+
+This file deals with the following version of the hydra game: each head of the hydra is
+labelled by an element in a type `α`, and when you cut off one head with label `a`, it
+grows back an arbitrary but finite number of heads, all labelled by elements smaller than
+`a` with respect to a well-founded relation `r` on `α`. We show that no matter how (in
+what order) you choose cut off the heads, the game always terminates, i.e. all heads will
+eventually be cut off (but of course it can last arbitrarily long, i.e. takes an
+arbitrary finite number of steps).
+
+This result is stated as the well-foundedness of the `cut_expand` relation defined in
+this file: we model the heads of the hydra as a multiset of elements of `α`, and the
+valid "moves" of the game are modelled by the relation `cut_expand r` on `multiset α`:
+`cut_expand r s' s` is true iff `s'` is obtained by removing one head `a ∈ s` and
+adding back an arbitrary multiset `t` of heads such that all `a' ∈ t` satisfy `r a' a`.
+
+To prove this theorem, we follow the proof by Peter LeFanu Lumsdaine at
+https://mathoverflow.net/a/229084/3332, and introduce the notion of `fibration` of relations, and
+a special addition of relations `game_add` that is used to define addition of games in
+combinatorial game theory.
+
+TODO: formalize the relations corresponding to more powerful (e.g. Kirby–Paris and Buchholz)
+hydras, and prove their well-foundedness.
+-/
+
+namespace relation
+
+variables {α β : Type*}
+
+section two_rels
+variables (rα : α → α → Prop) (rβ : β → β → Prop) (f : α → β)
```

Comment:
There's no point in having `rα` and `rβ` on this `variables` block if they're only used explicitly once.

Reply:
I don't understand this comment; pretty much every result below mentions `rα` and `rβ`, and more than one have them as explicit arguments (certainly for `game_add` and `fibration`). Having them as `variables` save the hassle of writing down their types every time, although I sometimes need to write `variables {rα rβ}` or `(rα rβ)` to switch between explicit and implicit. If I don't have them as `variables` and must include them in every declaration, then I must specify their types in every declaration except `acc.of_fibration`.